# ML-07 — Baseline Action Score and Top-20 Review

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/malaikasaleem944/malaika_flyrank/blob/main/work/notebooks/w04_baseline_score.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My rule and its reason codes

Rule:

Rank opportunities using a baseline score based on staleness and search volume. Higher staleness increases priority because stale pages may need refreshing, while higher search volume increases the potential value of the opportunity. The highest-scoring opportunities receive the corresponding action label.

Reason codes:

HIGH_STALENESS — the opportunity receives priority because its staleness signal is high.

HIGH_VOLUME — the opportunity receives priority because its search volume is high.

## 2. Build the ranked queue (writes the CSV)

*Code the score, rank everything, write work/outputs/baseline_action_score.csv.*

In [10]:
import duckdb
import pandas as pd

con = duckdb.connect()

print("DuckDB connected successfully")

DuckDB connected successfully


In [11]:
import os
from getpass import getpass

HF_TOKEN = os.getenv("HF_TOKEN")

if not HF_TOKEN:
    HF_TOKEN = getpass("Enter your Hugging Face token: ")

print("HF token loaded:", bool(HF_TOKEN))

HF token loaded: True


In [12]:
import duckdb

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [13]:
tables = con.sql(f"""
    SELECT *
    FROM glob('{REL}/*.parquet')
    LIMIT 20
""").df()

tables

,file
0,hf://datasets/FlyRank/internship-warehouse/dim...
1,hf://datasets/FlyRank/internship-warehouse/dim...
2,hf://datasets/FlyRank/internship-warehouse/fac...
3,hf://datasets/FlyRank/internship-warehouse/fac...


In [14]:
for file in tables["file"]:
    print(file)

hf://datasets/FlyRank/internship-warehouse/dim_clients.parquet
hf://datasets/FlyRank/internship-warehouse/dim_content.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_daily_performance_sample.parquet
hf://datasets/FlyRank/internship-warehouse/fact_content_query_90d.parquet


In [15]:
import os
import duckdb
from getpass import getpass

HF_TOKEN = os.getenv("HF_TOKEN")

if not HF_TOKEN:
    HF_TOKEN = getpass("Enter your Hugging Face token: ")

con = duckdb.connect()

con.execute(f"""
CREATE OR REPLACE SECRET hf (
    TYPE huggingface,
    TOKEN '{HF_TOKEN}'
)
""")

REL = "hf://datasets/FlyRank/internship-warehouse"

print("Hugging Face connection configured")

Hugging Face connection configured


In [16]:
print("DuckDB version:", duckdb.__version__)

DuckDB version: 1.5.5


In [17]:
import urllib.request

req = urllib.request.Request(
    "https://huggingface.co/api/whoami-v2",
    headers={"Authorization": f"Bearer {HF_TOKEN}"}
)

try:
    with urllib.request.urlopen(req) as response:
        print("Hugging Face authentication: OK")
except Exception as e:
    print("Hugging Face authentication failed")
    print(e)

Hugging Face authentication failed
HTTP Error 401: Unauthorized


In [18]:
import os
from getpass import getpass

HF_TOKEN = os.getenv("HF_TOKEN")

if not HF_TOKEN:
    HF_TOKEN = getpass("Enter your Hugging Face token: ")

print("HF token loaded:", bool(HF_TOKEN))

HF token loaded: True


In [19]:
import urllib.request

req = urllib.request.Request(
    "https://huggingface.co/api/whoami-v2",
    headers={"Authorization": f"Bearer {HF_TOKEN}"}
)

try:
    with urllib.request.urlopen(req) as response:
        print("Hugging Face authentication: OK")
except Exception as e:
    print("Hugging Face authentication failed")
    print(e)

Hugging Face authentication failed
HTTP Error 401: Unauthorized


In [20]:
print("Token exists:", bool(HF_TOKEN))
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)
print("Token prefix:", HF_TOKEN[:4] if HF_TOKEN else "NONE")

Token exists: True
Token length: 28
Token prefix: YOUR


In [21]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:4] if HF_TOKEN else "NONE")

Token exists: True
Token prefix: YOUR


In [22]:
HF_TOKEN = None
print("Old placeholder removed:", HF_TOKEN is None)

Old placeholder removed: True


In [23]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:3] if HF_TOKEN else "NONE")

Token exists: True
Token prefix: YOU


In [24]:
import urllib.request

req = urllib.request.Request(
    "https://huggingface.co/api/whoami-v2",
    headers={"Authorization": f"Bearer {HF_TOKEN}"}
)

try:
    with urllib.request.urlopen(req) as response:
        print("Hugging Face authentication: OK")
except Exception as e:
    print("Hugging Face authentication failed:", e)

Hugging Face authentication failed: HTTP Error 401: Unauthorized


In [25]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:3] if HF_TOKEN else "NONE")
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)

Token exists: True
Token prefix: YOU
Token length: 28


In [26]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:3] if HF_TOKEN else "NONE")
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)

Token exists: True
Token prefix: YOU
Token length: 28


In [27]:
from dotenv import load_dotenv
import os

load_dotenv()

HF_TOKEN = os.getenv("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:3] if HF_TOKEN else "NONE")
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)

Token exists: True
Token prefix: YOU
Token length: 28


In [28]:
import os

os.environ.pop("HF_TOKEN", None)

print("Old HF_TOKEN removed")

Old HF_TOKEN removed


In [29]:
import os

HF_TOKEN = os.environ.get("HF_TOKEN")

print("Token exists:", bool(HF_TOKEN))
print("Token prefix:", HF_TOKEN[:3] if HF_TOKEN else "NONE")
print("Token length:", len(HF_TOKEN) if HF_TOKEN else 0)

Token exists: False
Token prefix: NONE
Token length: 0


In [30]:
import pandas as pd

df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

print("Rows:", len(df))
print("Columns:", len(df.columns))
print(df.columns.tolist())

Rows: 30000
Columns: 44
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


In [31]:
# Create a simple baseline action score.
# Higher score = higher priority for action.

df["score"] = 0

# High impressions = enough search volume to justify an action
df.loc[df["impressions_90d"] >= 1000, "score"] += 2

# Position 5-20 = close enough to page 1 / strong ranking opportunity
df.loc[
    (df["avg_position"] >= 5) & (df["avg_position"] <= 20),
    "score"
] += 2

# Position 1-4 is already strong, so do not prioritize it
df.loc[df["avg_position"] == 0, "score"] = 0

# One reason code
df["reason_code"] = "NOT_PRIORITIZED"

df.loc[
    (df["impressions_90d"] >= 1000) &
    (df["avg_position"] >= 5) &
    (df["avg_position"] <= 20),
    "reason_code"
] = "HIGH_VOLUME_WEAK_POSITION"

# Action label
df["action"] = "MONITOR"

df.loc[
    df["reason_code"] == "HIGH_VOLUME_WEAK_POSITION",
    "action"
] = "QUICK_WIN"

In [32]:
# Rank all content items by score.
# Tie-break with impressions so higher-volume items come first.

ranked = df.sort_values(
    ["score", "impressions_90d"],
    ascending=[False, False]
).reset_index(drop=True)

ranked["rank"] = ranked.index + 1

print("Total rows ranked:", len(ranked))
print(ranked[[
    "rank",
    "content_id",
    "client_id",
    "score",
    "reason_code",
    "action",
    "impressions_90d",
    "avg_position"
]].head(10))

Total rows ranked: 30000
   rank            content_id          client_id  score  \
0     1  content_aaef01a50def  client_19581e27de      4   
1     2  content_db5989a78dd3  client_4e07408562      4   
2     3  content_cb112fce36be  client_19581e27de      4   
3     4  content_36ff89c8214e  client_19581e27de      4   
4     5  content_aa4baf490b43  client_349c41201b      4   
5     6  content_c84a0ab98e90  client_f369cb89fc      4   
6     7  content_c21024970297  client_19581e27de      4   
7     8  content_cea79ef51519  client_f369cb89fc      4   
8     9  content_c8e9d6ab9013  client_19581e27de      4   
9    10  content_d17681677e69  client_19581e27de      4   

                 reason_code     action  impressions_90d  avg_position  
0  HIGH_VOLUME_WEAK_POSITION  QUICK_WIN           517109           5.4  
1  HIGH_VOLUME_WEAK_POSITION  QUICK_WIN           345111           5.4  
2  HIGH_VOLUME_WEAK_POSITION  QUICK_WIN           309910           5.6  
3  HIGH_VOLUME_WEAK_POSITION  QUI

In [33]:
from pathlib import Path

OUTPUT_PATH = Path("../../work/outputs/baseline_action_score.csv")
OUTPUT_PATH.parent.mkdir(parents=True, exist_ok=True)

ranked.to_csv(OUTPUT_PATH, index=False)

print("CSV written to:", OUTPUT_PATH)
print("Rows written:", len(ranked))

CSV written to: ..\..\work\outputs\baseline_action_score.csv
Rows written: 30000


In [34]:
top10 = ranked.head(10)[[
    "rank",
    "content_id",
    "client_id",
    "score",
    "reason_code",
    "action",
    "impressions_90d",
    "avg_position"
]]

display(top10)

,rank,content_id,client_id,score,reason_code,action,impressions_90d,avg_position
0,1,content_aaef01a50def,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,517109,5.4
1,2,content_db5989a78dd3,client_4e07408562,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,345111,5.4
2,3,content_cb112fce36be,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,309910,5.6
3,4,content_36ff89c8214e,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,295097,7.3
4,5,content_aa4baf490b43,client_349c41201b,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,256290,5.9
5,6,content_c84a0ab98e90,client_f369cb89fc,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,223271,7.8
6,7,content_c21024970297,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,211366,5.1
7,8,content_cea79ef51519,client_f369cb89fc,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,208798,5.2
8,9,content_c8e9d6ab9013,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,208678,9.7
9,10,content_d17681677e69,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,201584,5.8


## 3. Top-20 review

*For each of the top 20: action, reason code, confidence note, and what would make it wrong.*

In [35]:
top20 = ranked.head(20)[[
    "rank",
    "content_id",
    "client_id",
    "score",
    "reason_code",
    "action",
    "impressions_90d",
    "avg_position"
]]

display(top20)

,rank,content_id,client_id,score,reason_code,action,impressions_90d,avg_position
0,1,content_aaef01a50def,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,517109,5.4
1,2,content_db5989a78dd3,client_4e07408562,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,345111,5.4
2,3,content_cb112fce36be,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,309910,5.6
3,4,content_36ff89c8214e,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,295097,7.3
4,5,content_aa4baf490b43,client_349c41201b,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,256290,5.9
5,6,content_c84a0ab98e90,client_f369cb89fc,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,223271,7.8
6,7,content_c21024970297,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,211366,5.1
7,8,content_cea79ef51519,client_f369cb89fc,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,208798,5.2
8,9,content_c8e9d6ab9013,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,208678,9.7
9,10,content_d17681677e69,client_19581e27de,4,HIGH_VOLUME_WEAK_POSITION,QUICK_WIN,201584,5.8


### Top-20 Review

1. **content_aaef01a50def** — Action: QUICK_WIN. Why: It has 517,109 impressions and an average position of 5.4, giving it very high volume and a clear ranking opportunity. What would make it wrong: The impressions may come from low-value or poorly matched queries.

2. **content_db5989a78dd3** — Action: QUICK_WIN. Why: It has 345,111 impressions and an average position of 5.4, so it combines high search volume with room to improve ranking. What would make it wrong: The searches may not represent valuable user intent.

3. **content_cb112fce36be** — Action: QUICK_WIN. Why: It has 309,910 impressions and an average position of 5.6, making it a strong high-volume ranking opportunity. What would make it wrong: Higher ranking may not improve useful traffic if query relevance is weak.

4. **content_36ff89c8214e** — Action: QUICK_WIN. Why: It has 295,097 impressions and an average position of 7.3, indicating substantial volume with room for improvement. What would make it wrong: The page may not satisfy the intent behind those searches.

5. **content_aa4baf490b43** — Action: QUICK_WIN. Why: It has 256,290 impressions and an average position of 5.9, combining strong volume with ranking opportunity. What would make it wrong: Users may not engage with the page even if its ranking improves.

6. **content_c84a0ab98e90** — Action: QUICK_WIN. Why: It has 223,271 impressions and an average position of 7.8, giving it meaningful volume and room to move upward. What would make it wrong: The search demand may be broad or poorly aligned with the content.

7. **content_c21024970297** — Action: QUICK_WIN. Why: It has 211,366 impressions and an average position of 5.1, making it a high-volume page with ranking potential. What would make it wrong: The average position may hide poor performance for the most valuable queries.

8. **content_cea79ef51519** — Action: QUICK_WIN. Why: It has 208,798 impressions and an average position of 5.2, showing substantial volume and an opportunity to improve visibility. What would make it wrong: The page may rank for many queries that are not commercially or strategically important.

9. **content_c8e9d6ab9013** — Action: QUICK_WIN. Why: It has 208,678 impressions and an average position of 9.7, giving it high volume but more room for ranking improvement. What would make it wrong: The impressions may be generated by low-value queries.

10. **content_d17681677e69** — Action: QUICK_WIN. Why: It has 201,584 impressions and an average position of 5.8, indicating strong volume with ranking opportunity. What would make it wrong: Improving position may not produce useful results if the content does not match search intent.

11. **content_a7427266c305** — Action: QUICK_WIN. Why: It has 201,111 impressions and an average position of 5.7, making it another high-volume ranking opportunity. What would make it wrong: The search volume may not translate into valuable traffic.

12. **content_2db251d1a841** — Action: QUICK_WIN. Why: It has 198,671 impressions and an average position of 5.6, combining substantial visibility with room for improvement. What would make it wrong: Query relevance or content quality may limit the benefit of a higher ranking.

13. **content_bf7bff5d0756** — Action: QUICK_WIN. Why: It has 197,199 impressions and an average position of 6.8, indicating strong volume and an improvable ranking. What would make it wrong: The page may not be the best destination for the searches producing those impressions.

14. **content_9463d30d5826** — Action: QUICK_WIN. Why: It has 192,478 impressions and an average position of 5.7, giving it high search volume and ranking opportunity. What would make it wrong: A higher ranking may not improve meaningful outcomes if search intent is weak.

15. **content_c5063073d048** — Action: QUICK_WIN. Why: It has 192,205 impressions and an average position of 12.5, so it has high volume and especially substantial room for ranking improvement. What would make it wrong: Its lower position may reflect poor relevance or content quality that cannot be fixed by simple optimization.

16. **content_33b4dceecad1** — Action: QUICK_WIN. Why: It has 181,574 impressions and an average position of 6.2, making it a high-volume page with an opportunity to improve ranking. What would make it wrong: The traffic opportunity may be overstated if the queries are low value.

17. **content_ea7cea4a8fd9** — Action: QUICK_WIN. Why: It has 169,602 impressions and an average position of 5.4, combining meaningful volume with ranking opportunity. What would make it wrong: Improving its average position may not improve useful traffic if intent alignment is poor.

18. **content_eb366e871254** — Action: QUICK_WIN. Why: It has 168,060 impressions and an average position of 16.6, giving it high volume but a relatively weak ranking and therefore a potentially large opportunity. What would make it wrong: The weak ranking may indicate that the content is not competitive or relevant enough to justify optimization.

19. **content_62ed76850efc** — Action: QUICK_WIN. Why: It has 167,858 impressions and an average position of 5.1, giving it strong volume and a clear ranking opportunity. What would make it wrong: The impressions may not correspond to valuable or relevant searches.

20. **content_14af58873e4f** — Action: QUICK_WIN. Why: It has 163,479 impressions and an average position of 6.5, combining high volume with room for ranking improvement. What would make it wrong: Users may not find the content useful even if its search position improves.

## 4. Weak picks + leakage check

*Which picks look wrong and why? Confirm no product flags or future windows leaked in.*

In [36]:
weak_picks = ranked.tail(5)[[
    "rank",
    "content_id",
    "score",
    "reason_code",
    "action",
    "impressions_90d",
    "avg_position"
]]

display(weak_picks)

,rank,content_id,score,reason_code,action,impressions_90d,avg_position
29995,29996,content_321383d0e9cf,0,NOT_PRIORITIZED,MONITOR,1,0.0
29996,29997,content_6bca373d8751,0,NOT_PRIORITIZED,MONITOR,1,0.0
29997,29998,content_7d12e9e7a4a5,0,NOT_PRIORITIZED,MONITOR,1,0.0
29998,29999,content_23dce6a656e6,0,NOT_PRIORITIZED,MONITOR,1,0.0
29999,30000,content_c322796023c8,0,NOT_PRIORITIZED,MONITOR,1,0.0


### Weak Picks

The weakest picks are low-priority because they have almost no observed search volume and an average position of 0, which means no usable position data rather than a ranking of zero.

1. **content_321383d0e9cf** — Score: 0, Action: MONITOR. Why: It has only 1 impression and avg_position = 0, so there is not enough evidence to prioritize it. What would make this wrong: Additional reliable search data could show that the content has meaningful demand.

2. **content_6bca373d8751** — Score: 0, Action: MONITOR. Why: It has only 1 impression and no usable position data, so the rule correctly gives it low priority. What would make this wrong: A later valid measurement could reveal meaningful search demand or ranking opportunity.

3. **content_7d12e9e7a4a5** — Score: 0, Action: MONITOR. Why: It has only 1 impression and avg_position = 0, providing too little evidence for a quick-win action. What would make this wrong: More reliable observations could change its priority.

4. **content_23dce6a656e6** — Score: 0, Action: MONITOR. Why: It has only 1 impression and no usable ranking position, so there is insufficient evidence for prioritization. What would make this wrong: Future valid measurements could show meaningful search volume and a ranking opportunity.

5. **content_c322796023c8** — Score: 0, Action: MONITOR. Why: It has only 1 impression and avg_position = 0, so the rule does not have enough evidence to recommend action. What would make this wrong: Additional reliable data could show that the content deserves attention.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it.

- [x] The notebook runs top to bottom with no errors (Runtime → Run all).

- [x] No client names, URLs, or private queries anywhere.

- [x] My claims use careful words: observed, measured, directional, decision-support.

- [x] Committed to my repo under `work/notebooks/` — then submit my repo URL on the card. Done.